design/plugin.md

```diff
@@ -0,0 +1,131 @@
+path = "bin/dcos-hello"
+description = "Say Hello"
+
+[[commands]]
+name = "hallo"
+path = "bin/dcos-hallo"
+description = "Say Hallo auf Deutsch"
+```
+
+### package.json
+
+The presence of this file indicates that the plugin was installed through the universe (eg. dcos package install kafka --cli).
+
+## dcos plugin add
+
+This new subcommand installs a package or a plugin for the corresponding DC/OS cluster. It is not a package manager, it only takes a path to a file ( ./…) or a URL (http(s)://), downloads the file, and puts it in the correct directory.
+
+It takes the following flags:
+
+- --update: We don’t want to overwrite, if a user adds a plugin that already exists. Instead, an error message should be printed stating that the plugin is already installed. If the user passes the path or URL of an already existing plugin but also adds the flag --update (or -u) flag, we will accept the reinstallation of the plugin.
+
+We define two types of plugins that can be installed: **zip** plugins and **bin** plugins.
+
+### ZIP plugins
+
+ZIP plugins are simply unpacked into the UUID directory of the attached cluster, such as:
+
+```
+$ tree ~/.dcos
+    \- clusters
+        \- 31781309-DD09-4B73-AEA6-05B4A2F6B6FC
+            \- subcommands
+                \- helloworld (plugin folder, please see “Plugin Structure” section)
+```
+
+A ZIP plugin is extracted into the env subfolder. The content of a ZIP plugin might be for example :
+
+```
+  bin/
+    dcos-hello
+    dcos-goodbye
+  plugin.toml
+```
+
+### BIN plugins
+
+BIN plugins (which only consist of a single binary executable file) are copied into the same location as ZIP plugins. A barebone plugin structure gets created, which only contains an env/bin folder with the single binary. At installation time a BIN plugin doesn’t contain a plugin.toml file. Please see the “Migration from legacy plugins or BIN plugins” section below.
+
+While BIN plugins are not deprecated, they do not support the same set of features as ZIP plugins (eg. the yet-to-come bash autocompletion feature).
+
+## dcos plugin remove
+
+As its name suggests, this command removes plugins (their name being given as arguments) locally by removing their directory from the filesystem.
+
+## dcos plugin list
+
+This command lists installed plugins, with their name, version, description and the subcommands they provide. It also accepts a --json flag.
+
+## Calling a plugin
+
+When a plugin is called it will have the top level command name as its second argument regardless of whether the plugin only has one executable.
+
+For example, if a `dcos-hello` executable in a plugin has a `hello` top level command, it would be called like `env/bin/dcos-hello hello [...]`.
+
+Executables contained within plugins are called synchronously from the CLI by spawning a new child process. The CLI waits for the child process to complete. The CLI makes available its own executable path to the child via an ENV variable `DCOS_CLI_EXECUTABLE_PATH`.
```

Comment:
The description of `DCOS_CLI_EXECUTABLE_PATH` probably belongs in the list below.

Reply:
The list below is for env variables which are only set when a cluster is attached.

Reply:
Ah I see now.